### PR TITLE
 PAAS-13878 Close unused sockets, port config

### PR DESCRIPTION
**Pull request by ZepZep**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 # Parameters for build
 params = {
     'name': name,
-    'version': '2.1.13',
+    'version': '2.1.17',
```
Comment (pbenas) [inline setup.py]: I guess you could name it just 2.1.14

File: smoker/server/plugins/__init__.py [hunk 2]
```diff
@@ -461,6 +463,14 @@ def __init__(self, name, queue, params, forced=False):
 
     def run(self):
         setproctitle.setproctitle('smokerd plugin %s' % self.plugin_name)
+
+        # close unnecessary open sockets created while cloning the subprocess
```
Comment (pbenas) [inline smoker/server/plugins/__init__.py]: not exactly "created", it's more like "cloned on fork"

File: smoker/server/plugins/__init__.py [hunk 3]
```diff
@@ -461,6 +463,14 @@ def __init__(self, name, queue, params, forced=False):
...
+        open_sockets = filter(lambda x: type(x) == socket._socketobject, gc.get_objects())
+        for s in open_sockets:
+            if s.type != 2:
```
Comment (pbenas) [inline smoker/server/plugins/__init__.py]: please explain what the magic constant of 2 stands for in a comment

File: smoker/server/plugins/__init__.py [hunk 4]
```diff
@@ -461,6 +463,14 @@ def __init__(self, name, queue, params, forced=False):
 
     def run(self):
         setproctitle.setproctitle('smokerd plugin %s' % self.plugin_name)
+
+        # close unnecessary open sockets created while cloning the subprocess
+        open_sockets = filter(lambda x: type(x) == socket._socketobject, gc.get_objects())
+        for s in open_sockets:
```
Comment (pbenas) [inline smoker/server/plugins/__init__.py]: "s" is not a nice variable name, I'd rather go with something like "socket"

File: smoker/server/plugins/__init__.py [hunk 5]
```diff
@@ -461,6 +463,14 @@ def __init__(self, name, queue, params, forced=False):
...
+            if s.type != 2:
+                #lg.info("Closing socket %s in plugin %s" % (s.getsockname(), self.plugin_name))
```
Comment (pbenas) [inline smoker/server/plugins/__init__.py]: please drop the commented out line

File: smoker/server/plugins/__init__.py [hunk 6]
```diff
@@ -461,6 +463,14 @@ def __init__(self, name, queue, params, forced=False):
...
+                s.close()
+
```
Comment (pbenas) [inline smoker/server/plugins/__init__.py]: maybe this code would be worth a nice separate function 'close_unecessary_sockets', similar to the 'drop_privileges'

File: smoker/client/__init__.py [hunk 7]
```diff
@@ -290,15 +292,18 @@ class Host(object):
 
     _result  = None
 
-    def __init__(self, address):
+    def __init__(self, address, config):
```
Comment (pbenas) [inline smoker/client/__init__.py]: you probably don't need to pass the whole config here, you could use an advantage of default parameter value like 
```
def __init__(self, address, port=8086)
```
and move the logic of checking whether port was supplied in the config to some calling class/function

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: smoker.spec [hunk 8]
```diff
@@ -1,8 +1,8 @@
 %global with_check 0
 
 Name:		smoker
-Version:	2.1.13
-Release:	1%{?dist}
+Version:	2.1.17
+Release:	2%{?dist}
```
Comment (pbenas) [inline smoker.spec]: I wouldn't bump the release here, the `make version` target ignores it and it's completely invisible on the python egg versioning level.

File: smoker/server/plugins/__init__.py [hunk 9]
```diff
@@ -738,6 +742,14 @@ def get_param(self, name, default=None):
         except KeyError:
             return default
 
+    def close_unnecessary_sockets(self):
+        """close unnecessary open sockets cloned on fork"""
+        open_sockets = filter(lambda x: type(x) == socket._socketobject, gc.get_objects())
+        for cur_socket in open_sockets:
+            # close all sockets except udp (type 2) - udp is used by Syslog
+            if cur_socket.type != 2:
```
Comment (pbenas) [inline smoker/server/plugins/__init__.py]: I'd rather see the logic inverted, close all TCP and leave everything else.
Also, no need to use the magic constant of two, the `socket` module offeres the constants predefined https://docs.python.org/2/library/socket.html#socket.SOCK_STREAM

File: setup.py (hunk 1 above)
Comment (pbenas) [reply setup.py]: really no need to go to 17, I can delete the scratch build out of the testing repo for you

File: smoker/client/__init__.py [hunk 10]
```diff
@@ -19,16 +19,20 @@ class Client(object):
     """
     hosts = []
 
-    def __init__(self, hosts):
+    def __init__(self, hosts, config):
```
Comment (pbenas) [inline smoker/client/__init__.py]: how about passing only the port number, together with the default parameter value?
I mean something like:
```python
def __init__(self, hosts, port=8086)
```

File: smoker/client/__init__.py [hunk 11]
```diff
@@ -19,16 +19,20 @@ class Client(object):
...
         """
         Create Host objects
         """
         assert isinstance(hosts, list), "Parameter hosts should be list"
 
+        default_port = 8086
+        if config and 'bind_port' in config and isinstance(config['bind_port'], int):
```
Comment (pbenas) [inline smoker/client/__init__.py]: and then you could put this logic to the `_host_discovery`

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/smoker-merge-compliance-check/1/ : SUCCESS in 14s

**Pull request merged**